include/sys/onoff.h

```diff
@@ -0,0 +1,533 @@
+ */
+int onoff_service_init(struct onoff_service *srv,
+		       onoff_service_transition_fn start,
+		       onoff_service_transition_fn stop,
+		       onoff_service_transition_fn reset,
+		       u32_t flags);
+
+/**
+ * @brief Flag fields used to specify on-off client behavior.
+ *
+ * These flags control whether calls to onoff_service_request() and
+ * onoff_service_release() are synchronous or asynchronous, and for
+ * asynchronous operations how the operation result is communicated to
+ * the client.
+ */
+enum onoff_client_flags {
+	/* Indicates operations are to proceed only if the initiating
+	 * function can be executed in a synchronous path.
+	 *
+	 * I.e. the operation will complete before the function
+	 * returns, though the operation may fail.
+	 */
+	ONOFF_CLIENT_NOTIFY_NO_WAIT     = 0,
+
+	/* Indicates operation completion notification is through a
+	 * k_poll_signal reference.
+	 */
+	ONOFF_CLIENT_NOTIFY_SIGNAL      = 1,
+
+	/* Indicates operation completion notification is through a
+	 * callback.
+	 */
+	ONOFF_CLIENT_NOTIFY_CALLBACK    = 2,
+
+	/* This and higher bits reserved for internal use. */
+	ONOFF_CLIENT_INTERNAL_BASE      = BIT(2),
+};
+
+/* Forward declaration */
+struct onoff_client;
+
+/**
+ * @brief Signature used to notify an on-off service client of the
+ * completion of a transition.
+ *
+ * These functions may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-callable and non-suspendable.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param res the result of the transition.  Expected values are
+ * service-specific, but the value shall be non-negative if the
+ * operation succeeded, and negative if the operation failed.
+ */
+typedef void (*onoff_client_callback)(struct onoff_service *srv,
+				      struct onoff_client *cli,
+				      int res);
+
+/**
+ * @brief State associated with a client of an on-off service.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. onoff_client_init_signal()) to
+ * configure them.
+ *
+ * Control of the object content transfers to the service provider
+ * when a pointer to the object is passed to any on-off service
+ * function.  While the service provider controls the object the
+ * client must not change any object fields.  Control reverts to the
+ * client:
+ * * if the call to the service API returns an error; or
+ * * if the call to the service API succeeds for a no-wait operation; otherwise
+ * * when operation completion is posted (signalled or callback invoked).
+ *
+ * Only the result field is intended for use by clients, and it is
+ * available for inspection only after control reverts to the client.
+ */
+struct onoff_client {
+	/* Links the client into the set of waiting service users. */
+	sys_snode_t node;
+
+	union {
+		/* Pointer to signal used to notify client.
```

Comment:
Whatever operation is currently active using this structure: request, release, or reset.

Reply:
haven't you considered having operation type argument (req, rel, reset)? This is known to the service when notifying client. Client need to have additional service state to handle that properly.
e.g.
```
void notification(struct onoff_service *srv,
				      struct onoff_client *cli,
				      int res)
{
  client_owner = container_of(cli);
  if ((client_state== SRV_REQUESTED) && (res < 0)) {
     client_state = RESETTING;
    onoff_service_reset(...);
  } else {
  ...
  }
}
```
As you can see, this handling gets complicated.

I'm also missing `user_data` in that callback. Need to use `container_of` and it is not always the case that it can be used to get entity owning the client.

Reply:
No, I hadn't, because the client provides the structure initialized for the specific operation.  I would expect something like:

```
static void notified_requested(struct onoff_service *srv,
                               struct onoff_client *cli,
                               int res)
{
  if (res < 0) {
     cli->async.callback = notify_reset;
     int rc = onoff_service_reset(srv, cli);
     if (rc < 0) {
        do_something();
     }
  } else {
     do_something_else();
  }
}
```
I think the reaction to each specific transition type is likely to be completely distinct.  Until there's evidence that a common callback provides significant value I'd rather leave it as is.

I'll add `user_data`, though in all cases I've come up with so far `CONTAINER_OF()` is sufficient.  It would be nice if we could make that a standard feature of all asynchronous callback APIs, but when I tried a year ago I got significant push-back.

Reply:
> I'll add user_data, though in all cases I've come up with so far CONTAINER_OF() is sufficient.

It is not sufficient if client owner is have pointer to client:
```
struct client cli;
struct owner {
  struct client *cli;
}

struct client cli;
const struct owner o = {.cli = &cli};
```

Regarding changing the callback. You can of course do that but that requires callback element to be in ram. However, in general i prefer to keep separate constant parts of the module from volatile parts to reduce ram usage (`struct shell` as the example) and even this is not the case now in client structure i would favor parameter in the callback.

Reply:
The `onoff_client` structure that holds the notification details must be in RAM, because it has to be linkable into a list.  So you'll always be able to change the callback (my example code was "wrong" because it did it directly; it should have used `onoff_client_init_callback()`.)

Reply:
i was not suggesting that it can be done here. I was rather pointing out that this approach completely closes this option.